server/service/src/plugin/validation.rs

```diff
@@ -0,0 +1,289 @@
+
+use pem::Pem;
+use rsa::pkcs8::{DecodePrivateKey, DecodePublicKey};
+use rsa::pss::{Signature, SigningKey, VerifyingKey};
+use rsa::sha2::Sha256;
+use rsa::signature::{RandomizedSigner, SignatureEncoding, Verifier};
+use rsa::{RsaPrivateKey, RsaPublicKey};
+
+use walkdir::WalkDir;
+use x509_parser::prelude::{FromDer, X509Certificate};
+
+use super::manifest::{create_manifest, Manifest, ManifestSignatureInfo};
+use super::{
+    CERTIFICATE_TAG, MANIFEST_FILE, MANIFEST_SIGNATURE_FILE, PLUGIN_CERT_DIR, PLUGIN_FILE_DIR,
+    PRIVATE_KEY_TAG, SHA256_NAME, SIGNATURE_TAG, VERIFICATION_ALGO_PSS,
+};
+
+#[derive(Clone)]
+pub struct ValidatedPlugin {
+    /// Modification date of the manifest file, e.g. to check if plugin has been modified and needs
+    /// to be validated again.
+    pub manifest_datetime: SystemTime,
+    pub manifest: Manifest,
+}
+
+#[derive(Clone)]
+pub struct ValidatedPluginBucket {
+    /// plugin base directory
+    plugin_dir: PathBuf,
+    /// Dir containing all trusted public plugin certs
+    trusted_cert_path: PathBuf,
+    /// Mapping the absolute plugin to a ValidatedPlugin
+    manifests: HashMap<String, ValidatedPlugin>,
+}
+
+impl ValidatedPluginBucket {
+    pub fn new(base_dir: &Option<String>) -> anyhow::Result<Self> {
+        let plugin_dir = match base_dir {
+            Some(base_dir) => PathBuf::from_str(base_dir)?.join(PLUGIN_FILE_DIR),
+            None => PathBuf::from_str(PLUGIN_FILE_DIR)?,
+        };
+        let trusted_cert_path = match base_dir {
+            Some(base_dir) => PathBuf::from_str(base_dir)?.join(PLUGIN_CERT_DIR),
+            None => PathBuf::from_str(PLUGIN_CERT_DIR)?,
+        };
+
+        Ok(ValidatedPluginBucket {
+            plugin_dir,
+            trusted_cert_path,
+            manifests: HashMap::new(),
+        })
+    }
+
+    pub fn validate_plugin(&mut self, path: &PathBuf) -> anyhow::Result<ValidatedPlugin> {
+        let path = path.canonicalize()?;
+        let path_string = path.as_os_str().to_string_lossy().to_string();
+        if let Some(plugin) = self.manifests.get(&path_string) {
+            let metadata = File::open(path.join(MANIFEST_FILE))?.metadata()?;
+            if metadata.modified()? == plugin.manifest_datetime {
+                return Ok(plugin.clone());
+            }
+        };
+        self.reload()?;
+        if let Some(plugin) = self.manifests.get(&path_string) {
+            return Ok(plugin.clone());
+        }
+        Err(anyhow::Error::msg(format!(
+            "Failed to validate plugin: {:?}",
+            path
+        )))
+    }
+
+    fn reload(&mut self) -> anyhow::Result<()> {
+        let certs = load_trusted_certs_from_dir(&self.trusted_cert_path)?;
+
+        self.manifests.clear();
+        let walker = WalkDir::new(&self.plugin_dir).max_depth(1);
+        for entry in walker {
+            let entry = entry?;
+
+            // be conservative and take manifest the timestamp before validating the plugin
```

Comment:
🤔  Unsure what the intention was - something like this?

```suggestion
            // be conservative and check the manifest before validating the plugin
```

Reply:
Thanks I clearified it and rearranged the code a bit to make it more clear:

```rust
            // Be conservative and record the manifest timestamp before validating the plugin.
            // For example, when the plugin changes while validating it, the older timestamp would
            // trigger a reload when fetching a plugin (in validate_plugin()).
            let metadata = File::open(&manifest_path)?.metadata()?;
            let manifest_datetime = metadata.modified()?;
```